server-metrics/src/main/scala/org/http4s/server/metrics/Metrics.scala

```diff
@@ -2,102 +2,203 @@ package org.http4s
+      put_req = m.timer(s"${prefix}.put-requests"),
+      head_req = m.timer(s"${prefix}.head-requests"),
+      move_req = m.timer(s"${prefix}.move-requests"),
+      options_req = m.timer(s"${prefix}.options-requests"),
+      trace_req = m.timer(s"${prefix}.trace-requests"),
+      connect_req = m.timer(s"${prefix}.connect-requests"),
+      delete_req = m.timer(s"${prefix}.delete-requests"),
+      other_req = m.timer(s"${prefix}.other-requests"),
+      total_req = m.timer(s"${prefix}.requests")
+    )
 
-      total_req.update(elapsed, TimeUnit.NANOSECONDS)
-      active_requests.dec()
-    }
+    val serviceMetrics = ServiceMetrics(generalServiceMetrics, requestTimers, responseTimers)
 
-    def onFinish(method: Method, start: Long)(
-        r: Either[Throwable, Option[Response[F]]]): Either[Throwable, Option[Response[F]]] = {
-      val elapsed = System.nanoTime() - start
-
-      r.map { r =>
-          headers_times.update(System.nanoTime() - start, TimeUnit.NANOSECONDS)
-          val code = r.fold(Status.NotFound)(_.status).code
-
-          def capture(body: EntityBody[F]): EntityBody[F] =
-            body
-              .onFinalize {
-                F.delay {
-                  generalMetrics(method, elapsed)
-                  if (code < 200) resp1xx.update(elapsed, TimeUnit.NANOSECONDS)
-                  else if (code < 300) resp2xx.update(elapsed, TimeUnit.NANOSECONDS)
-                  else if (code < 400) resp3xx.update(elapsed, TimeUnit.NANOSECONDS)
-                  else if (code < 500) resp4xx.update(elapsed, TimeUnit.NANOSECONDS)
-                  else resp5xx.update(elapsed, TimeUnit.NANOSECONDS)
-                }
-              }
-              .handleErrorWith { cause =>
-                abnormal_terminations.update(elapsed, TimeUnit.NANOSECONDS)
-                Stream.raiseError(cause)
-              }
-          r.map(resp => resp.copy(body = capture(resp.body)))
-        }
-        .leftMap { e =>
-          generalMetrics(method, elapsed)
-          resp5xx.update(elapsed, TimeUnit.NANOSECONDS)
-          service_errors.update(elapsed, TimeUnit.NANOSECONDS)
-          e
-        }
-    }
+    Kleisli(metricsService[F](serviceMetrics, service)(_))
+  }
 
-    Kleisli { req =>
-      val now = System.nanoTime()
-      active_requests.inc()
-      OptionT {
+  private def metricsService[F[_]: Sync](serviceMetrics: ServiceMetrics, service: HttpService[F])(
+      req: Request[F]): OptionT[F, Response[F]] = {
+    val method = req.method
+    for {
+      now <- OptionT.liftF[F, Long](Sync[F].delay(System.nanoTime()))
+      _ <- OptionT.liftF[F, Unit](
+        Sync[F].delay(serviceMetrics.generalMetrics.active_requests.inc()))
+      resp <- OptionT {
         service(req).value.attempt
-          .flatMap(onFinish(req.method, now)(_)
-            .fold[F[Option[Response[F]]]](F.raiseError, F.pure))
+          .flatMap(metricsServiceHandler(method, now, serviceMetrics, _))
```

Comment:
This block seems redundant to `liftF` multiple times, then wrap the last one. 

You can instead make the comprehension only in terms of `F` and wrap the whole thing. It would be better since `liftF` causes a `map(Some(_))`.